fdbclient/NativeAPI.actor.cpp

```diff
@@ -2192,6 +2217,131 @@ ACTOR Future<Void> watchValue(Future<Version> version, Key key, Optional<Value>
+			
+			metadata->watchPromise.send(ver);
+			metadata->watchFutureSS.cancel();
+		}
+
+		if (valSS == value) { // val_3 == val_2 (storage server value matches value passed into the function -> new watch)
+			metadata = makeReference<WatchMetadata>(value, ver, info, tags);
+			cx->setWatchMetadata(key, metadata);
+
+			metadata->watchFutureSS = watchStorageServerResp(key, cx);
+		}
+
+		if (valSS != value) return Void(); // if val_3 != val_2
+
+		wait(success(metadata->watchPromise.getFuture())); // val_3 == val_2
+
+		return Void();
+	}
+	catch(Error& e) {
+		state Error err = e;
+		wait(tr.onError(e));
+		throw err;
+	}
+}
+
+Future<Void> getWatchFuture(Version ver, Key key, Optional<Value> value, Database cx,
+                              TransactionInfo info, TagSet tags) {
+	Reference<WatchMetadata> metadata = cx->getWatchMetadata(key);
+
+	if (!metadata.isValid()) { // case 1: key not in map
+		metadata = makeReference<WatchMetadata>(value, ver, info, tags);
+		cx->setWatchMetadata(key, metadata);
+
+		metadata->watchFutureSS = watchStorageServerResp(key, cx);
+		return success(metadata->watchPromise.getFuture());
+	} 
+	else if (metadata->value == value) { // case 2: val_1 == val_2 (received watch with same value as key already in the map so just update)
+		if (ver > metadata->version) {
+			metadata->version = ver;
+			metadata->info = info;
+			metadata->tags = tags;
+		}
+		
+		return success(metadata->watchPromise.getFuture());
+	} else if(ver > metadata->version) { // case 3: val_1 != val_2 && version_2 > version_1 (recived watch with different value and a higher version so recreate in SS)
+		TEST(true); // Setting a watch that has a different value than the one in the map but a higher version (newer)
+		cx->deleteWatchMetadata(key);
+
+		metadata->watchPromise.send(ver);
+		metadata->watchFutureSS.cancel();
+
+		metadata = makeReference<WatchMetadata>(value, ver, info, tags);
+		cx->setWatchMetadata(key, metadata);
+
+		metadata->watchFutureSS = watchStorageServerResp(key, cx);
+
+		return success(metadata->watchPromise.getFuture());
+	} else if (metadata->version == ver) { // case 5: val_1 != val_2 && version_1 == version_2 (recived watch with different value but same version)
+		TEST(true); // Setting a watch which has a different value than the one in the map but the same version
+		return sameVersionDiffValue(ver, key, value, cx, info, tags);
+	}
+	TEST(true); // Setting a watch which has a different value than the one in the map but a lower version (older)
+	// case 4: val_1 != val_2 && version_2 < version_1
+	return Void();
+}
+
+ACTOR Future<Void> watchValueMap(Future<Version> version, Key key, Optional<Value> value, Database cx,
+                              TransactionInfo info, TagSet tags) {
+	state Version ver = wait(version);
+	try {
+		Future<Void> watchFuture = getWatchFuture(ver, key, value, cx, info, tags);
+		wait(watchFuture);
+		return Void();
+	} catch (Error & e) {
+		throw e;
```

Comment:
Actually I think this `try-catch` block is necessary because if the user cancels the watch future and it goes out of scope it will throw `error_code_operation_cancelled` so we need to throw in order to not run into any weird issues